File: day3/puzzle2.py
x = [0,0]
y = [0,0]
houses = {}
i = 0
with open('day3.txt') as f:
 while True:
  address = ':'.join([str(x[i]),str(y[i])])
  presents = 0
  if houses.has_key(address):
   presents = houses[address]
  houses[address] = presents + 1
  c = f.read(1)
  if not c:
   break
  if c == '^':
   y[i] = y[i] - 1
  if c == 'v':
   y[i] = y[i] + 1
  if c == '>':
   x[i] = x[i] + 1
  if c == '<':
   x[i] = x[i] - 1
  i = (i + 1) & 1
print len(houses)

File: 2015/03/puzzle2.js
var fs = require('fs');

fs.readFile('day3.txt', 'utf-8', function (err, inp) {
 if (err) throw err;

 var x = [0,0], y = [0,0], i = 0, houses = {};
 var dirs = { '^': [0,-1], 'v': [0,1], '>': [1,0], '<': [-1,0] };
 inp.trim().split("").forEach(function (c) {
  var address = x[i] + ":" + y[i];
  houses[address] = (houses[address] || 0) + 1;
  var dir = dirs[c];
  x[i] += dir[0]; y[i] += dir[1]; i ^= 1;
 });
 console.log(Object.keys(houses).length);
});
